### PR TITLE
`Exercise Chat`: Implement native function calling agent

### PR DESCRIPTION
**Pull request by kaancayli**
This PR implements the exercise chat as a native tool calling agent. For now only OpenAI tool calling is supported.

More details will be added.

## Summary by CodeRabbit

- **New Features**
	- Enhanced message conversion functionality to support new message types and roles, including tool messages.
	- Introduced the `ExerciseChatAgentPipeline` for improved interactions regarding exercise-related queries.
	- Added flexibility in the chat pipeline to handle different variants and improve response handling.
	- Implemented new methods for binding tools within language models, enhancing interaction capabilities.
	- Updated prompts for improved interaction with students during exercise submissions and feedback.
	- Introduced `PyrisEventDTO` for flexible event handling.
	- Added a new `IRIS_CHAT_EXERCISE_AGENT_MESSAGE` enum member to represent pipeline states.

- **Bug Fixes**
	- Improved robustness in message handling and state management to prevent invalid transitions.

- **Documentation**
	- Updated prompts and guidelines for the AI tutor to enhance educational interactions.

- **Chores**
	- Added new dependencies to the project for enhanced language processing capabilities.
	- Updated existing dependencies to newer versions for improved performance and security.

### CODE REVIEW

File: app/pipeline/chat/course_chat_pipeline.py [hunk 1]
```diff
@@ -121,6 +125,7 @@
             :param dto: The pipeline execution data transfer object
             :param kwargs: The keyword arguments
         """
+        print(dto.model_dump_json(indent=4))
```
[the remark this answers is not in the document]
Comment (yassinsws) [reply app/pipeline/chat/course_chat_pipeline.py:128]: I think we are using logging now instead of printing

**Review by yassinsws — APPROVED** [review]
Code looks good only one minor nitpick

File: app/llm/external/openai_chat.py [hunk 2]
```diff
@@ -141,14 +218,21 @@
...
         raise Exception(f"Failed to get response from OpenAI after {retries} retries")
 
+    def bind_tools(
+        self,
+        tools: Sequence[
+            Union[Dict[str, Any], Type[LegacyBaseModel], Callable, BaseTool]
+        ],
+    ):
+        self.tools = tools
+
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply app/llm/external/openai_chat.py:235]: Thanks I applied the suggested change

File: app/domain/event/pyris_event_dto.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from typing import TypeVar, Generic
+
+from pydantic import Field, BaseModel
+
+T = TypeVar("T")
+
+
+class PyrisEventDTO(BaseModel, Generic[T]):
+    event_type: str = Field(None, alias="eventType")
+    event: T = Field(None, alias="event")
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply app/domain/event/pyris_event_dto.py:10]: I will apply this.

File: app/llm/external/openai_chat.py [hunk 4]
```diff
@@ -77,30 +115,50 @@
...
+                        "name": tc.function.name,
+                        "arguments": tc.function.arguments,
+                    },
+                )
+                for tc in message.tool_calls
+            ],
+            contents=[TextMessageContentDTO(textContent="")],
+            sendAt=datetime.now(),
+            token_usage=tokens,
+        )
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply app/llm/external/openai_chat.py:140]: In those places, it's only checked if the contents array is empty, messages inside the array can have a content with empty string.

File: app/llm/external/openai_chat.py [hunk 5]
```diff
@@ -113,20 +167,43 @@ def chat(
...
+                            max_tokens=arguments.max_tokens,
+                            tools=self.tools,
+                        )
+                    else:
+                        response = client.chat.completions.create(
+                            model=self.model,
+                            messages=messages,
+                            temperature=arguments.temperature,
+                            max_tokens=arguments.max_tokens,
+                        )
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply app/llm/external/openai_chat.py:206]: Thanks, I will apply this.

File: app/llm/request_handler/basic_request_handler.py [hunk 6]
```diff
@@ -33,3 +37,11 @@ def chat(
...
         llm = self.llm_manager.get_llm_by_id(self.model_id)
         return llm.embed(text)
+
+    def bind_tools(
+        self,
+        tools: Sequence[Union[Dict[str, Any], Type[BaseModel], Callable, BaseTool]],
+    ) -> LanguageModel:
+        llm = self.llm_manager.get_llm_by_id(self.model_id)
+        llm.bind_tools(tools)
+        return llm
```
[the remark this answers is not in the document]
Comment (kaancayli) [reply app/llm/request_handler/basic_request_handler.py:47]: Thanks, I will apply this suggestion

**Review by yassinsws — APPROVED** [review]
Exercise Chaat pipeline works well with lecture chat pipeline good job Kaan

**Review by alexjoham — APPROVED** [review]
Code looks good to me! Also checked the token tracking, works as it should!

**Pull request merged**